### ISSUE TITLE
Save to SAF folder

### ISSUE THREAD

**Issue by Vito0912** [781flyingdutchman/background_downloader#425]
**Is your feature request related to a problem? Please describe.**  
I target clients using android versions above 30, which require SAF (Storage Access Framework) to save files outside the internal storage. Currently, files must be downloaded locally and then manually moved to an SAF/URI folder. I would like an option to download files directly to an SAF/URI folder.

**Describe the solution you'd like**  
Allow files to be downloaded directly to an SAF folder as described above.

**Describe alternatives you've considered**  
The current alternative is saving files to local storage and then moving them to an SAF folder.

**Additional context**  
Currently, there seems to be no way to achieve this. If this feature is already implemented/ there is a more simple way, I apologize for the oversight.

**Comment by 781flyingdutchman** [comment]
Thanks for raising this. Currently not available - I recently implemented the option to use mediaStore UIRs for uploads (see #398) but then realized that that is only a small part of the URI space in Android (see #415). Storage management on Android is complicated and fragmented, so I am not currently committing to implementing this. 

The currently suggested approach for what you describe is to download to a temporary location (not necessarily the temp directory, as that also has complexity in Android) and move ito your desired destination. You'l have to use another packlage to obtain the URI and move the file there.  Perhaps take a look at [this](https://github.com/781flyingdutchman/background_downloader/discussions/321#discussioncomment-11131913) which explains a new feature that would allow you to execute the move upon completion of the donwload/upload, even if the app is in the background.

**Comment by Vito0912** [comment]
Thank you for your response! 
I've already implemented this in my application and only wanted to bring it to your attention in case it could be considered for future updates. 
Currently, I’m using the package mentioned in (398), but I’d love to see this feature integrated down the line someday as it would bring away complexity and error handling from the clients. 
Regardless, thank you for creating such an amazing plugin!

**Comment by 781flyingdutchman** [comment]
Hi @Vito0912 I am looking into this. 

Three questions:
1. What package/approach do you use to get the URI of the directory that a user selects (which will then be used to download the file directly to that destination)? I looked at `file_picker` but that gives me a path, not the URI, and I really need the proper SAF URI for this
2. I looked into the file_picker package, and it returns a directory path. Have you tried using that path with the `FileDownloader` by setting `baseDirectory` to `BaseDirectory.root` and passing the path you got from the file picker? If that works, then we don't need the URI complication (but I am not sure you're supposed to translate an SAF URI to a file path and start writing to it)
3. Is your issue related to not being able to set a user-selected SAF destination as the target for the `FileDOwnloader` (which may be fixed if item 2 aboe works) or is the issue that the downloader first downloads to a temporary file and then copies it to the final destination?  If it's the latter, then perhaps I could create a configuration that changes that behavior

 Please let me know your thoughts on these. Thanks!

**Comment by Vito0912** [comment]
First of all, thanks for looking into this. I used SAF because it seemed like the only way to write to external storage in newer Android versions. If you know of a better way, that should probably be preferred, and SAF should not be used. I'm not very knowledgeable about Android specifics though.

Answers to your questions:

1. I use https://github.com/flutter-cavalry/saf_stream and saf_util. It returns the content://... URL / aka URI.

2. I initially tried your approach but received complaints that it did not work when setting outside the app's internal storage and it did indeed not work. I tested this by using the file_picker package to select a directory and then tried to write to it. It didn't work. I think the approach works for single files but not as a persistent writable directory. For other platforms, I use your BaseDirectory.root and the path returned by the picker idea. It just didn't work for Android. Maybe I made a mistake in handling permissions. I cannot and do not want to request access to the entire filesystem.

3. I can't write to the directory returned by the file selector, so I cannot download files directly into it. Instead, I use the downloader to download the file to the temp directory and then copy it using the `saf_stream` package. I tried providing the URI as a path with the root directory, but, as expected, it did not work.

**Comment by Vito0912** [comment]
>The file will be downloaded to a temp file like normal, and then copied into the destination defined by the URI + filename. 

This sounds good, although downloading it directly to the SAF would obviously be better, especially for large files like media files. I just want to point out that I currently do not need to fix this/I currently don't have a "problem", as I already have implement many workarounds to make it work (my downloader logic is messy because every platform requires its own small adjustments) and currently I am using the method to download to a temp folder and then moving to a SAF folder. Therefore currently I also don't really plan to change it that much (as it would mean that the fromUri only works on Android and the logic must be sperated anyways). I would "only" use it if it downloaded directly - But no stress or anything, just want to be honest about using it then. 
At the moment, I also have an issue with passing the SAF files to other libraries, so there is no rush at all.

Still, I believe it would be a nice benefit for others to use, and honestly, I am curious how they managed it, given that this restriction has existed in some Android versions for a while now. Starting with downloading it to the temp directory seems to work well for most cases.

Tldr; If I hadn’t implemented the temp solution myself, then yes, this would solve my problem.

**Comment by 781flyingdutchman** [comment]
Understood, no worries, it's very helpful to hear what you've tried. I tried the saf_utils pickDirectory method, and had some questions: 
1. How do you figure out what uri to use for the initial location, eg if you want it to open at the Movies directory? 
2. Regardless of whether I ask for persistence, the dialog asks me if I want to grant access now and in the future. That seems wrong to me if I don't ask for persistence. The dialog also comes up every time - doesn't seem to remember I already granted access. Is that your experience?
3. Likewise, whether I ask for read or write access, I can write to the uri regardless. 

The package uses the raw Android API so I guess this is how it is supposed to work, but I wanted to check if that is consistent with your experience?

**Comment by Vito0912** [comment]
I experience the same, yes

I believe it supports an initial path as a URI, but I'm not entirely sure. Since I only use it in one specific location, it wasn't particularly important to me where it opens.

I think you are more knowledgeable about Android-related topics. My guess would be that there’s only a persistent method. To be honest, I’m not completely sure what “non-persistent” means in this context—maybe it only lasts for a few days? If that’s the case, the label “in the future” might make sense.

Probably the same as above. It’s also possible this is an issue with the library itself, as it’s very small and relatively new. I just couldn't find a better alternative.

**Comment by 781flyingdutchman** [comment]
Thanks again. I do think there's an opportunity to streamline the "pick a destination and save it there" or "pick a file and upload it" across all platforms, without the need for platform-specific logic. I'd like to try - would you be comfortable sharing the code snippet of how you currently implement the logic such that it works across platforms (hide irrelevant/sensitive code of course)? It would help me understand the complexities and figure out how to abstract those away.

I'd be aiming for 1) a call to show a dialog, giving you a platform-agnostic result object that you can then use to 2) create an upload or download task without any further steps. Separately I'll implement an option to bypass the temp file and download straight to the destination.

I know I've asked a lot, but would really appreciate your help here!

**Comment by Vito0912** [comment]
No problem! I’ll link my project here. Just keep in mind that my downloader logic is a mess. 😄 Don't ask me what I did there, but it got worse and worse. I know it's definitely not the proper way to listen for updates/ do it.

### Code References:

#### 1. Code to Retrieve the Current Directory:
This is my code to retrieve the current directory:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/features/settings/platform_settings.dart#L123).

#### 2. File Saving Logic (When a Download Finishes): 
Lines **118–141** (*until 138 for the URI* logic) handle saving the downloaded files to the SAF storage from the temp storage:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/provider/downloader_provider.dart#L118).

#### 3. Download Logic:
Here is the main download logic to begin the process:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/provider/downloader_provider.dart#L388).

#### 4. `savePath` Definition:
Here is where the `savePath` is defined, which is appended:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/provider/downloader_provider.dart#L413).

#### 5. Overwriting the User-Set Download Path (Excluding Android):
If there is a user-set download path, the above path will be overwritten for non-Android systems. For Android I do not change the save path, but just change the BaseDir to temporary:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/provider/downloader_provider.dart#L422).

#### 6. Custom Path for Android:
If an Android user selects a custom path, the logic chooses to use the temporary storage instead of `BaseDirectory.root`. This behavior can be seen in the `else` statement (Else is for other platforms set to have a custom path [just the path returned by FilePicker].:  
[**View code here**](https://github.com/Vito0912/abs_flutter/blob/fe672863963a1ec30eb070d2baf7c288f15876b5/lib/provider/downloader_provider.dart#L434).

---

### Summary:

When an Android user selects a custom path, my logic uses the temporary storage (instead of the base directory). Then, I listen for the download finish event (see the second link above) to move the file out of temporary storage and into the SAF file.

As a result:
1. It temporarily takes **double the disk space**.  
2. It takes a **longer time** to copy large media files into the SAF location. Because I have a media app (even if it's just Audio) it takes some time. But this seems to be an issue with the used library generally, I guess

---

*itemId* and *episodeId* are just some uuids, which I use to generate the path.

**Comment by 781flyingdutchman** [comment]
Hi, just a status update and request for feedback.  I am working on implementing the following interface for working with Uris:
```
  /// Opens a directory picker dialog and returns the selected directory's URI.
  ///
  /// [startLocation] (optional) specifies a [SharedStorage] location to open the picker at.
  /// [startLocationUri] (optional) specifies a URI to open the picker at.
  /// Only one of [startLocation] or [startLocationUri] should be provided.
  /// [persistedUriPermission] (optional, defaults to `false`) indicates whether to take persisted URI permission
  /// for the selected directory, if the platform supports it.
  ///
  /// Returns the selected directory's URI, or `null` if the user canceled the operation.
  Future<Uri?> pickDirectory(
      {SharedStorage? startLocation,
      Uri? startLocationUri,
      bool persistedUriPermission = false});

  /// Opens a file picker dialog and returns a list of the selected files' URIs.
  ///
  /// [startLocation] (optional) specifies a [SharedStorage] location to open the picker at.
  /// [startLocationUri] (optional) specifies a URI to open the picker at.
  /// Only one of [startLocation] or [startLocationUri] should be provided.
  /// [allowedExtensions] (optional) specifies a list of file extensions to filter the picker by.
  /// [multipleAllowed] (optional, defaults to `false`) indicates whether to allow multiple file selection.
  /// [persistedUriPermission] (optional, defaults to `false`) indicates whether to take persisted URI permission
  /// for the selected files, if the platform supports it.
  ///
  /// Returns a list of the selected files' URIs, or `null` if the user canceled the operation.
  Future<List<Uri>?> pickFiles(
      {SharedStorage? startLocation,
      Uri? startLocationUri,
      List<String>? allowedExtensions,
      bool multipleAllowed = false,
      bool persistedUriPermission = false});

  /// Creates a new directory with the given name within the specified parent directory.
  ///
  /// [parentDirectoryUri] is the URI of the parent directory.
  /// [newDirectoryName] is the name of the new directory to create.
  /// [persistedUriPermission] (optional, defaults to `false`) indicates whether to take persisted URI permission
  /// for the created directory, if the platform supports it.
  ///
  /// Returns the URI of the newly created directory.
  Future<Uri> createDirectory(Uri parentDirectoryUri, String newDirectoryName,
      {bool persistedUriPermission = false});

  /// Retrieves the file data (bytes) for a given URI.
  ///
  /// [uri] is the URI of the file.
  ///
  /// Returns a [Uint8List] containing the file data, or `null` if an error occurred.
  Future<Uint8List?> getFileBytes(Uri uri);

  /// Deletes the file at the given URI.
  ///
  /// [uri] is the URI of the file to delete.
  ///
  /// Returns `true` if the file was deleted successfully, `false` otherwise.
  Future<bool> deleteFile(Uri uri);

  /// Opens the file at [uri] with the given [mimeType].
  ///
  /// Returns true if successful, false otherwise.
  Future<bool> openFile(Uri uri, {String mimeType});
```

This goes together with `DownloadTask.fromUri` and `UploadTask.fromUri`, and essentially switches the downloader/uploader to "Uri mode" instead of "filepath mode". Because Uris are platform-agnostic, you can now write code such as:
```
final directory = await FileDownload().uri.pickDirectory(); // lets user pick a directory
final downloadDir = await FileDownloader.uri.createDirectory(directory, 'newDirName'); // creates a subdir there
final task = DownloadTask.fromUri(uri: doownloadDir, url: 'https://example.com', filename: 'name');
final result = await FileDownloader().download(task); // downloads directly into downloadDir, skips temp file
final bytes = await FileDownloader().uri.openFile(result.task.fileUri); // opens the file
```

There are still some limitations and inconsistencies between platforms, but this should be much better.

Would love your feedback on this, and a question: how do you want to access the downloaded file? Can you always get a filePath on Android (eg if this is stored in external storage)?

**Comment by Vito0912** [comment]
I will provide more feedback tomorrow, but I wanted to address the question:  

> How do you want to access the downloaded file?  

I’m not sure, to be honest. Ideally, it would be as a `File`, but that doesn’t seem to be possible (at least I couldn’t find any package for it). Currently, I’m just using the returned URI. A `File` would definitely be better, as it would ensure consistency across all systems.  

> Can you always get a filePath on Android (e.g., if it’s stored in external storage)?  

As far as I know, only the URI exists. I think it’s possible to list files in a directory, so probably. But for my application, I just save them locally.

**Comment by 781flyingdutchman** [comment]
Yes, that is why I was asking. If the Uri is a content:// Uri on Android then the only way to access it is via that Uri (you cannot get a file path). In the proposed interface, if you set `persistedUriPermission` to true, the Uris you get will remain valid through device restarts, so effectively become a permanent location for the file that you can store in a database. In combination with the proposed `getFileBytes` method you can then get the bytes into Dart. Unfortunately you cannot pass an Android bytes stream to a Dart bytes stream, so you have to work with the bytes as a blob.

Desktop will work with file:// Uris that can also be translated into a path (though for consistency the same interface will work, eg you can get the fileBytes via the Uri). On iOS there is yet another mechanism for persistent Uris so you are back to having to use that in most cases. They key is that everywhere you should be able to use the Uri as a permanent locator for the file/directory.

The package implements the file pickers on iOS and Android (so no need to use the file_picker package) but it does not for desktop. If you want to use the same on desktop, then for those platforms you have to use the file_picker package and convert the file path that you get back into a Uri with `Uri.file(pathToDirectoryOrFile)` and use that Uri just like the mobile Uris in the downloader. While that's an extra step, there are two advantages: 1) file_picker package not needed when app is only mobile (vast majority of apps) and 2) still consistent interface with the downloader.

**Comment by Vito0912** [comment]
Ah sorry, I totally forgot. My bad.

Regarding the last comment, I am fine with it.

>This goes together with DownloadTask.fromUri and UploadTask.fromUri, and essentially switches the downloader/uploader to "Uri mode" instead of "filepath mode". Because Uris are platform-agnostic, you can now write code such as:

I don't know if you want to include this directly in your library, but maybe consider having a constructor for something like `fromDynamic` (My naming is very bad). This would use URI when URI is used and the normal file path for others.

**Comment by 781flyingdutchman** [comment]
This is now implemented on branch V9. It will stay there for a while until this becomes main, so until then you can try it out by changing your pubspec.yaml entry to:
```background_downloader:
    git:
      url: https://github.com/781flyingdutchman/background_downloader.git
      ref: V9
```

Make sure to read the Uri document at docs/URI.md to see how to use it. Notably, I have moved away from named constructors for the Uri - use `UriDownloadTask` or `UriUploadTask` and use the `FileDownloader().uri` object to access pickers etc.

**Comment by 781flyingdutchman** [comment]
@Vito0912 wondering if you've had a chance to explore this new approach, and if so if you have any feedback?

**Comment by 781flyingdutchman** [comment]
Implemented in V9.0.0

**Comment by Vito0912** [comment]
Sorry. I did not see the old message. No, I currently did not test it, but I assume it works great :) Thank you for your great work
